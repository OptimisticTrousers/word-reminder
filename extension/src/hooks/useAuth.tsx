// import { useMutation } from "@tanstack/react-query";
// import { useEffect, useState } from "react";
// import { useNavigate } from "react-router-dom";
// import { toast } from "react-toastify";

// const useAuth = () => {
//     const navigate = useNavigate();
//     // const [isAuthenticated, setIsAuthenticated] = useState(false);

//     // const {data, status, error, mutuate} = useMutation({mutationFn: () => {
//     //     return fetch(`${import.meta.env.VITE_API_DOMAIN}/current`)
//     // }})

//     const { data: loginData, status: loginStatus, error: loginError, mutate: loginMutate } = useMutation({
//         mutationFn: (formData) => {
//             return fetch(`${import.meta.env.VITE_API_DOMAIN}/login`, formData)
//         }, onSuccess: () => {
//             toast.success("You have successfully logged in!");
//             navigate("/")
//         }, onError: () => {
//             toast.error("There was an issue logging in!")
//         }
//     })

//     const { data: registerData, status: registerStatus, error: registerError, mutate: registerMutate } = useMutation({
//         mutationFn: (formData) => {
//             return fetch(`${import.meta.env.VITE_API_DOMAIN}/register`, formData)
//         }
//         , onSuccess: () => {
//             toast.success("You have successfully registered!");
//             navigate("/")
//         }, onError: () => {
//             toast.error("There was an issue registering!")
//         }
//     })

//     const { data: logoutData, status: logoutStatus, error: logoutError, mutate: logoutMutate } = useMutation({
//         mutationFn: (formData) => {
//             return fetch(`${import.meta.env.VITE_API_DOMAIN}/logout`, formData)
//         }
//         , onSuccess: () => {
//             toast.success("You have successfully logged out!");
//             navigate("/")
//         }, onError: () => {
//             toast.error("There was an issue logging out!")
//         }
//     })


//     useEffect(() => {

//     }, [])

//     return {
//         login: {
//             data: loginData,
//             status: loginStatus,
//             error: loginError,
//             mutate: loginMutate
//         }, register: {
//             data: registerData,
//             status: registerStatus,
//             error: registerError,
//             mutate: registerMutate
//         }, logout: {
//             data: logoutData, status: logoutStatus, error: logoutError, mutuate: logoutMutate
//         }
//     };
// }

// export default useAuth;