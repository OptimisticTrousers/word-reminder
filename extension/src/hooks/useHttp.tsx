type HttpMethods = "GET" | "POST" | "PUT" | "DELETE";

const useHttp = () => {
  const request = async (
    url: string,
    method: HttpMethods,
    options: RequestInit,
    body?: unknown
  ) => {
    const response = await fetch(url, {
      method,
      mode: "cors",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
      ...options,
      body: JSON.stringify(body),
    });

    const json = await response.json();

    return json;
  };

  const get = async (url: string, options: RequestInit = {}) => {
    return request(url, "GET", options);
  };

  const post = async (
    url: string,
    body: unknown,
    options: RequestInit = {}
  ) => {
    return request(url, "POST", options, body);
  };

  const put = async (url: string, body: unknown, options: RequestInit = {}) => {
    return request(url, "PUT", options, body);
  };

  const remove = async (url: string, options: RequestInit = {}) => {
    return request(url, "DELETE", options);
  };

  return { get, post, put, remove };
};

export default useHttp;
