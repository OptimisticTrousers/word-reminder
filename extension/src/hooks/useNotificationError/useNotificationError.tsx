import { useContext } from "react";
import { useNavigate } from "react-router-dom";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../context/Notification";
import { ErrorResponse } from "../../types";
import { AUTH_NOTIFICATION_MSGS } from "../../pages/Auth/constants";

export function useNotificationError() {
  const navigate = useNavigate();
  const { showNotification } = useContext(NotificationContext);

  function showNotificationError(response: ErrorResponse) {
    const message = response.json.message;
    if (message === "Incorrect email." || message === "Incorrect password.") {
      showNotification(
        NOTIFICATION_ACTIONS.ERROR,
        AUTH_NOTIFICATION_MSGS.incorrectCredentials()
      );
    } else if (message == "User is unauthenticated.") {
      showNotification(
        NOTIFICATION_ACTIONS.ERROR,
        AUTH_NOTIFICATION_MSGS.credentialsExpired()
      );
      navigate("/login");
    } else {
      showNotification(NOTIFICATION_ACTIONS.ERROR, response.json.message);
    }
  }

  return { showNotificationError };
}
