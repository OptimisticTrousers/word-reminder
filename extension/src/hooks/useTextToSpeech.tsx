import { useState, useEffect } from "react";

const useTextToSpeech = (text: string) => {
  const [utterance, setUtterance] = useState<SpeechSynthesisUtterance | null>(
    null
  );

  useEffect(() => {
    const synth = window.speechSynthesis;
    const u = new SpeechSynthesisUtterance(text);

    setUtterance(u);

    return () => {
      synth.cancel();
    };
  }, [text]);

  const handlePlay = () => {
    const synth = window.speechSynthesis;

    synth.speak(utterance!);
  };

  return { handlePlay };
};

export default useTextToSpeech;
