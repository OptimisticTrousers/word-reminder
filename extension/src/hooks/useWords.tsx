const useWords = () => {
  const { get, post } = useHttp();

  const {
    data: userWords,
    isSuccess,
    error: userWordsError,
  } = useQuery({
    queryKey: ["words"],
    queryFn: () => {
      return get(
        `${
          import.meta.env.VITE_API_DOMAIN
        }/users/665164760636f4834e053388/words`
      );
    },
  });
};
