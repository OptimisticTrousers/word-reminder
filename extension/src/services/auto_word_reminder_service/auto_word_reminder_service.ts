import { AutoWordReminderParams } from "common";

import { service } from "../service";

export const autoWordReminderService = (function (service) {
  const { get, post, put, remove, VITE_API_DOMAIN } = service;

  function getAutoWordReminder(userId: string) {
    return get({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders`,
      options: {
        credentials: "include",
      },
    });
  }

  function createAutoWordReminder({
    userId,
    body,
  }: {
    userId: string;
    body: Omit<AutoWordReminderParams, "user_id">;
  }) {
    return post({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function updateAutoWordReminder({
    userId,
    autoWordReminderId,
    body,
  }: {
    userId: string;
    autoWordReminderId: string;
    body: Omit<AutoWordReminderParams, "user_id">;
  }) {
    return put({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders/${autoWordReminderId}`,
      options: { body: JSON.stringify(body), credentials: "include" },
    });
  }

  function deleteAutoWordReminder({
    userId,
    autoWordReminderId,
  }: {
    userId: string;
    autoWordReminderId: string;
  }) {
    return remove({
      url: `${VITE_API_DOMAIN}/users/${userId}/autoWordReminders/${autoWordReminderId}`,
      options: { credentials: "include" },
    });
  }

  return {
    createAutoWordReminder,
    getAutoWordReminder,
    deleteAutoWordReminder,
    updateAutoWordReminder,
  };
})(service);
