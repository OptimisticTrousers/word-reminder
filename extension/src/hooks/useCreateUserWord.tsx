import { useMutation } from "@tanstack/react-query";
import useHttp from "./useHttp";

const useCreateUserWord = () => {
  

  return {
    createdUserWord,
    createdUserWordStatus,
    createdUserWordError,
    createdUserWordMutate,
  };
};

export default useCreateUserWord;
