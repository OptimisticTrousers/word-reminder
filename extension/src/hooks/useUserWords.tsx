import { useQuery } from "@tanstack/react-query";
import useAuth from "./useAuth";
import useHttp from "./useHttp";

interface Props {
  learned: boolean;
  sort: string;
  search: string;
}

const useUserWords = ({ learned, search, sort }: Props) => {
  const { get } = useHttp();
  const {
    state: { user },
  } = useAuth();

  const userId = user!._id;

  const query = new URLSearchParams();
  if (typeof learned === "boolean") {
    query.append("learned", learned.toString());
  }
  if (search) {
    query.append("search", search);
  }
  if (sort) {
    query.append("sort", sort);
  }

  const {
    data: userWords,
    status: userWordsStatus,
    error: userWordsError,
  } = useQuery({
    queryKey: [userId, "words", { learned, search, sort }],
    queryFn: () => {
      return get(
        `${import.meta.env.VITE_API_DOMAIN}/users/${userId}/words${
          query.size ? "?" : ""
        }${query}`
      );
    },
  });

  return { userWords, userWordsStatus, userWordsError };
};

export default useUserWords;
