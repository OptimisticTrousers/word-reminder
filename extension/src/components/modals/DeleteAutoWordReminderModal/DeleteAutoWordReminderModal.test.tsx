import { createRoutesStub, Outlet } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import userEvent from "@testing-library/user-event";

import { DeleteAutoWordReminderModal } from "./DeleteAutoWordReminderModal";
import * as hooks from "../../../hooks/useNotificationError";
import { autoWordReminderService } from "../../../services/auto_word_reminder_service/auto_word_reminder_service";

describe("DeleteAutoWordReminderModal component", () => {
  const json = {
    word: { id: "1" },
  };

  const testUser = {
    id: "1",
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders the form with alert and message", async () => {
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      autoWordReminderId: "1",
    };
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteAutoWordReminderModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);

    const { asFragment } = render(<Stub initialEntries={["/"]} />);
    const alert = screen.getByText(
      "Are you sure you want to delete this auto word reminder?"
    );
    const message = screen.getByText("You can't undo this action.");

    expect(alert).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the functions to delete a word", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      autoWordReminderId: "1",
    };
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteAutoWordReminderModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    const status = 200;
    const mockDeleteAutoWordReminder = vi
      .spyOn(autoWordReminderService, "deleteAutoWordReminder")
      .mockImplementation(async () => {
        return { json, status };
      });
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const user = userEvent.setup();

    render(<Stub initialEntries={["/"]} />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });
    await user.click(deleteButton);
    expect(mockInvalidateQueries).toHaveBeenCalledTimes(1);
    expect(mockInvalidateQueries).toHaveBeenCalledWith({
      queryKey: ["autoWordReminders"],
      exact: true,
    });
    expect(mockDeleteAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockDeleteAutoWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      autoWordReminderId: props.autoWordReminderId,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
  });

  it("calls the functions to show a notification error", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      autoWordReminderId: "1",
    };
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteAutoWordReminderModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    const status = 400;
    const message = "Error Message.";
    const mockDeleteAutoWordReminder = vi
      .spyOn(autoWordReminderService, "deleteAutoWordReminder")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });
    const mockShowNotificationError = vi.fn();
    vi.spyOn(hooks, "useNotificationError").mockImplementation(() => {
      return { showNotificationError: mockShowNotificationError };
    });
    const mockInvalidateQueries = vi.spyOn(queryClient, "invalidateQueries");
    const user = userEvent.setup();

    render(<Stub initialEntries={["/"]} />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });
    await user.click(deleteButton);
    expect(mockInvalidateQueries).not.toHaveBeenCalled();
    expect(mockDeleteAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockDeleteAutoWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      autoWordReminderId: props.autoWordReminderId,
    });
    expect(mockShowNotificationError).toHaveBeenCalledTimes(1);
    expect(mockShowNotificationError).toHaveBeenCalledWith({
      json: { message },
      status,
    });
    expect(mockToggleModal).toHaveBeenCalledTimes(1);
    expect(mockToggleModal).toHaveBeenCalledWith();
  });

  it("disables the delete button when the mutation is loading", async () => {
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      autoWordReminderId: "1",
    };
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteAutoWordReminderModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    const status = 200;
    const delay = 50;
    const mockDeleteAutoWordReminder = vi
      .spyOn(autoWordReminderService, "deleteAutoWordReminder")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json, status });
          }, delay);
        });
      });
    const user = userEvent.setup();

    render(<Stub initialEntries={["/"]} />);

    const deleteButton = screen.getByRole("button", { name: "Delete" });
    await user.click(deleteButton);

    expect(deleteButton).toBeDisabled();
    expect(mockDeleteAutoWordReminder).toHaveBeenCalledTimes(1);
    expect(mockDeleteAutoWordReminder).toHaveBeenCalledWith({
      userId: testUser.id,
      autoWordReminderId: props.autoWordReminderId,
    });
  });

  it("calls the functions to close a modal", async () => {
    const mockToggleModal = vi.fn();
    const props = {
      toggleModal: mockToggleModal,
      autoWordReminderId: "1",
    };
    const user = userEvent.setup();
    const queryClient = new QueryClient({
      defaultOptions: { queries: { retry: false } },
    });
    const Stub = createRoutesStub([
      {
        path: "/",
        Component: function () {
          return <Outlet context={{ user: testUser }} />;
        },
        children: [
          {
            path: "/",
            Component: function () {
              return (
                <QueryClientProvider client={queryClient}>
                  <DeleteAutoWordReminderModal {...props} />;
                </QueryClientProvider>
              );
            },
          },
        ],
      },
    ]);
    render(<Stub initialEntries={["/"]} />);

    const cancelButton = screen.getByRole("button", { name: "Cancel" });
    await user.click(cancelButton);

    expect(props.toggleModal).toHaveBeenCalledTimes(1);
    expect(props.toggleModal).toHaveBeenCalledWith();
  });
});
