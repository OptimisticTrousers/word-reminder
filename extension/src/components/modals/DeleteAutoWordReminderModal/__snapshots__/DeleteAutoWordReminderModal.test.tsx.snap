// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeleteAutoWordReminderModal component > renders the form with alert and message 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="modal-title"
    aria-modal="true"
    class="_modal_811945"
    role="dialog"
  >
    <div
      class="scale-down _modal__container_811945"
    >
      <header
        class="_modal__header_811945"
      >
        <button
          aria-label="Close modal"
          class="_modal__button_811945"
        >
          <svg
            class="lucide lucide-square-x _modal__icon_811945"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="3"
            />
            <path
              d="m15 9-6 6"
            />
            <path
              d="m9 9 6 6"
            />
          </svg>
        </button>
        <div
          class="_modal__title-bar_811945"
          id="modal-title"
        >
          <h2
            class="_modal__title_811945"
          >
            Delete Auto Word Reminder
          </h2>
        </div>
      </header>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="_modal_c90c1e"
      >
        <p
          class="_modal__alert_c90c1e"
        >
          Are you sure you want to delete this auto word reminder?
        </p>
        <p
          class="_modal__message_c90c1e"
        >
          You can't undo this action.
        </p>
        <div
          class="_modal__buttons_c90c1e"
        >
          <button
            class="_modal__button_c90c1e _modal__button--cancel_c90c1e"
          >
            Cancel
          </button>
          <button
            class="_modal__button_c90c1e _modal__button--delete_c90c1e"
            type="submit"
          >
            Delete
          </button>
        </div>
      </form>
    </div>
  </div>
  ;
</DocumentFragment>
`;
